# Variational Autoencoder (VAE) for EEG Data Augmentation

### 1. EEG Data Processing

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from pathlib import Path 
import mne

from sklearn.model_selection import GroupKFold, cross_validate
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.metrics import (
    accuracy_score,
    f1_score,
    roc_auc_score,
    confusion_matrix, 
    ConfusionMatrixDisplay
)
import tensorflow as tf
from tensorflow.keras import layers, models

2026-04-18 17:06:05.583919: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 AVX512F AVX512_VNNI FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.


In [2]:
def find_project_root():
    p = Path.cwd()
    for parent in [p, *p.parents]:
        if (parent / ".git").exists():
            return parent
    raise FileNotFoundError("Project root (with .git) not found")

PROJECT_ROOT = find_project_root()
RAW_DATA_DIR = PROJECT_ROOT / "data" / "raw" / "nm000114"
PROCESSED_DIR = PROJECT_ROOT / "data" / "processed_features"
RESULTS_DIR = PROJECT_ROOT / "results"

print("PROJECT_ROOT:", PROJECT_ROOT)
print("RAW_DATA_DIR:", RAW_DATA_DIR)
print("PROCESSED_DIR:", PROCESSED_DIR)
print("RESULTS_DIR:", RESULTS_DIR)

PROJECT_ROOT: /Users/trinityho/chem_277b/MSSE-277b-final-project-
RAW_DATA_DIR: /Users/trinityho/chem_277b/MSSE-277b-final-project-/data/raw/nm000114
PROCESSED_DIR: /Users/trinityho/chem_277b/MSSE-277b-final-project-/data/processed_features
RESULTS_DIR: /Users/trinityho/chem_277b/MSSE-277b-final-project-/results


In [3]:
def get_eeg_file(subject_id: str, condition: str):
    """ Get the path to the EEG file for a given subject and condition.
    """
    return RAW_DATA_DIR / subject_id / "eeg" / f"{subject_id}_task-{condition}_eeg.edf"

def infer_label_from_subject(subject_id: str) -> int:
    """
    Infer the label (0 for healthy control, 1 for MDD) from the subject ID.
    """
    subject_upper = subject_id.upper()
    if "HS" in subject_upper:
        return 0
    elif "MDD" in subject_upper:
        return 1 
    else: 
        raise ValueError(f"Could not infer label from subject ID: {subject_id}")
    
def parse_condition_from_filename(filepath: Path) -> str:
    """
    Parse the condition (eyesClosed, eyesOpen, P300) from the filename.
    """
    name = filepath.name 
    if "task-eyesClosed" in name:
        return "eyesClosed"
    elif "task-eyesOpen" in name:
        return "eyesOpen"
    elif "task-P300" in name:
        return "P300"
    else:
        raise ValueError(f"Could not parse condition from filename: {name}")
        
edf_files =sorted(RAW_DATA_DIR.glob("sub*/eeg/*.edf"))
        
rows = []
for filepath in edf_files:
    subject_id = filepath.parts[-3] 
    rows.append({
        "patient_id": subject_id,
        "recording_id": filepath.stem,
        "label": infer_label_from_subject(subject_id),
        "condition": parse_condition_from_filename(filepath),
        "filepath": str(filepath),
    })


# Create a DataFrame from the metadata
metadata_df = pd.DataFrame(rows)
metadata_df.head()

,patient_id,recording_id,label,condition,filepath
0,sub-HS1,sub-HS1_task-P300_eeg,0,P300,/Users/trinityho/chem_277b/MSSE-277b-final-pro...
1,sub-HS1,sub-HS1_task-eyesClosed_eeg,0,eyesClosed,/Users/trinityho/chem_277b/MSSE-277b-final-pro...
2,sub-HS1,sub-HS1_task-eyesOpen_eeg,0,eyesOpen,/Users/trinityho/chem_277b/MSSE-277b-final-pro...
3,sub-HS10,sub-HS10_task-P300_eeg,0,P300,/Users/trinityho/chem_277b/MSSE-277b-final-pro...
4,sub-HS10,sub-HS10_task-eyesClosed_eeg,0,eyesClosed,/Users/trinityho/chem_277b/MSSE-277b-final-pro...


In [4]:
print("Shape:", metadata_df.shape)
print("\nUnique patients:", metadata_df["patient_id"].nunique())
print("\nClass counts:")
print(metadata_df["label"].value_counts())
print("\nCondition counts:")
print(metadata_df["condition"].value_counts())
print("\nRecordings per patient:")
print(metadata_df.groupby("patient_id").size().value_counts())

metadata_path = PROCESSED_DIR / "metadata.csv"
metadata_df.to_csv(metadata_path, index=False)
print("\nMetadata saved to:", metadata_path)

Shape: (181, 5)

Unique patients: 64

Class counts:
label
1    95
0    86
Name: count, dtype: int64

Condition counts:
condition
eyesOpen      62
P300          61
eyesClosed    58
Name: count, dtype: int64

Recordings per patient:
3    52
2    10
4     1
1     1
Name: count, dtype: int64

Metadata saved to: /Users/trinityho/chem_277b/MSSE-277b-final-project-/data/processed_features/metadata.csv


In [5]:
COMMON_CHANNELS = [
'EEG Fp1-LE', 'EEG F3-LE', 'EEG C3-LE', 'EEG P3-LE', 'EEG O1-LE',
 'EEG F7-LE', 'EEG T3-LE', 'EEG T5-LE', 'EEG Fz-LE', 'EEG Fp2-LE', 
 'EEG F4-LE', 'EEG C4-LE', 'EEG P4-LE', 'EEG O2-LE', 'EEG F8-LE', 
 'EEG T4-LE', 'EEG T6-LE', 'EEG Cz-LE', 'EEG Pz-LE', 'EEG A2-A1'
]

print("Number of common channels:", len(COMMON_CHANNELS))

Number of common channels: 20


In [6]:
def extract_band_power(raw, COMMON_CHANNELS) -> np.ndarray:
    """
    Extract relative band-power features for each EEG channel in the raw MNE object.
    Returns a feature vector.
    """

    data_raw = raw.copy()
    data_raw.pick(COMMON_CHANNELS)
    
    # get data 
    data = data_raw.get_data()      # shape: (n_channels, n_time points)
    sfreq = data_raw.info['sfreq']  # sampling frequency (e.g. 256 Hz)

    # Define frequency bands 
    bands = {
        "delta": (1, 4),     # deep sleep
        "theta": (4, 8),     # drowsiness
        "alpha": (8, 13),    # relaxed 
        "beta":  (13, 30)    # active thinking 
    }

    all_features = []

    # Loop through each channel (ch) and convert signal to frequency domain using Welch's method to estimate power spectral density (psd)
    for ch in data:        
        psd, freqs = mne.time_frequency.psd_array_welch(
            ch,
            sfreq=sfreq,
            fmin=1,
            fmax=30,
            verbose=False
        )

        total_power = psd.sum()
        
        band_features = [
            (
                psd[(freqs >= fmin) & (freqs <= fmax)].mean()
            / total_power                                      # normalize by total power to get relative power in each band
            if total_power > 0 else 0                          # handle case where total power is zero to avoid division by zero
            )                           
            for (fmin, fmax) in bands.values()
            ]
    
        all_features.append(band_features)

    return np.nan_to_num(np.array(all_features).flatten())

Build Feature Matrix

In [7]:
# Build feature matrix 

records = [
    (
        extract_band_power(
            mne.io.read_raw_edf(row["filepath"], preload=True, verbose=False),
            COMMON_CHANNELS
        ),
        row["label"],
        row["patient_id"]
    )
    for _, row in metadata_df.iterrows()
]

X = np.vstack([r[0] for r in records])
y = np.array([r[1] for r in records])
groups = np.array([r[2] for r in records])

print("X.shape:", X.shape)
print("y.shape:", y.shape)
print("Groups shape:", groups.shape)

X.shape: (181, 80)
y.shape: (181,)
Groups shape: (181,)


# VAE for Each Condition (P300, Eyes Open, Eyes Closed)

In [165]:
# VAE components 
class Sampling(layers.Layer):
    def call(self, inputs):
        z_mean, z_log_var = inputs
        epsilon = tf.random.normal(shape=tf.shape(z_mean))
        return z_mean + tf.exp(0.5 * z_log_var) * epsilon

def build_encoder(input_dim, latent_dim):
    inputs = layers.Input(shape=(input_dim,))
    x = layers.Dense(64, activation="relu")(inputs)
    x = layers.Dense(32, activation="relu")(x)
    z_mean = layers.Dense(latent_dim)(x)
    z_log_var = layers.Dense(latent_dim)(x)
    return models.Model(inputs, [z_mean, z_log_var])

def build_decoder(output_dim, latent_dim):
    inputs = layers.Input(shape=(latent_dim,))
    x = layers.Dense(32, activation="relu")(inputs)
    x = layers.Dense(64, activation="relu")(x)
    outputs = layers.Dense(output_dim)(x)
    return models.Model(inputs, outputs)

def compute_loss(x, x_recon, z_mean, z_log_var):
    x = tf.cast(x, tf.float32)
    x_recon = tf.cast(x_recon, tf.float32)
    
    recon = tf.reduce_mean(tf.reduce_sum(tf.square(x - x_recon), axis=1))
    
    kl = -0.5 * tf.reduce_mean(
        tf.reduce_sum(1 + z_log_var - tf.square(z_mean) - tf.exp(z_log_var), axis=1)
    )
    return recon + kl


In [176]:
# train VAE
all_losses = []

def train_vae(X_data, latent_dim=8, epochs=15):
    scaler = StandardScaler()
    X_scaled = scaler.fit_transform(X_data).astype("float32")

    input_dim = X_scaled.shape[1]

    encoder = build_encoder(input_dim, latent_dim)
    decoder = build_decoder(input_dim, latent_dim)

    optimizer = tf.keras.optimizers.Adam()

    dataset = tf.data.Dataset.from_tensor_slices(X_scaled.astype("float32")).batch(64)

    epoch_losses = [] # track VAE training loss

    for epoch in range(epochs):
        batch_losses = []

        for batch in dataset:
            with tf.GradientTape() as tape:

                z_mean, z_log_var = encoder(batch)
                z = Sampling()([z_mean, z_log_var])
                x_recon = decoder(z)

                loss = compute_loss(batch, x_recon, z_mean, z_log_var)

            grads = tape.gradient(
                loss,
                encoder.trainable_weights + decoder.trainable_weights
            )

            optimizer.apply_gradients(
                zip(grads, encoder.trainable_weights + decoder.trainable_weights)
            )

            batch_losses.append(loss.numpy())

        # average loss for this epoch
        avg_loss = np.mean(batch_losses)
        epoch_losses.append(avg_loss)

        if (epoch + 1) % 10 == 0 or epoch == 0:
            print(f"Epoch {epoch+1}/{epochs} - Loss: {avg_loss:.4f}")

    return encoder, decoder, scaler, epoch_losses



In [177]:
# Augmentation of training sets
conditions = ["P300", "eyesOpen", "eyesClosed"]

def augment_training_data(X_train, y_train, metadata_train, vae_models, latent_dim=8, n_fake=20):

    X_aug = [X_train]
    y_aug = [y_train]

    for cond in conditions:

        mask = metadata_train["condition"].values == cond

        if mask.sum() < 10:
            continue

        decoder = vae_models[cond]["decoder"]
        scaler = vae_models[cond]["scaler"]

        for label in [0, 1]:

            z = np.random.normal(size=(n_fake, latent_dim)).astype("float32")

            X_fake_scaled = decoder(z, training=False)
            X_fake = scaler.inverse_transform(X_fake_scaled)

            y_fake = np.full(n_fake, label)

            X_aug.append(X_fake)
            y_aug.append(y_fake)

    return np.vstack(X_aug), np.hstack(y_aug)

# Training Loop for Combined Augmented Data

In [178]:
# Apply GroupKFold
all_vae_losses = []
gkf = GroupKFold(n_splits=5)

acc_real = []
acc_aug = []

f1_real = []
f1_aug = []

for fold, (train_idx, test_idx) in enumerate(gkf.split(X, y, groups)):
    print(f"\nFold {fold+1}")

    X_train, X_test = X[train_idx], X[test_idx]
    y_train, y_test = y[train_idx], y[test_idx]
    metadata_train = metadata_df.iloc[train_idx]

    # train VAE
    vae_models = {}
    vae_losses = {}

    for cond in conditions:

        mask = metadata_train["condition"].values == cond
        X_cond = X_train[mask]

        if len(X_cond) < 10:
            continue

        _, decoder, scaler, losses = train_vae(
            X_cond,
            latent_dim=8,
            epochs=30
        )

        vae_models[cond] = {
            "decoder": decoder,
            "scaler": scaler
        }
        
        vae_losses[cond] = losses
    
    all_vae_losses.append(vae_losses)

    # --- Baseline Model ---
    clf = Pipeline([
    ("scaler", StandardScaler()),
    ("logreg", LogisticRegression(max_iter=1000))])
    clf.fit(X_train, y_train)

    y_pred = clf.predict(X_test)
    acc_real.append(accuracy_score(y_test, y_pred))
    f1_real.append(f1_score(y_test, y_pred))

    # --- Augmentation ---
    X_aug, y_aug = augment_training_data(
        X_train,
        y_train,
        metadata_train,
        vae_models
    )

    clf_aug = Pipeline([
    ("scaler", StandardScaler()),
    ("logreg", LogisticRegression(max_iter=1000))])
    clf_aug.fit(X_aug, y_aug)

    y_pred_aug = clf_aug.predict(X_test)
    acc_aug.append(accuracy_score(y_test, y_pred_aug))
    f1_aug.append(f1_score(y_test, y_pred_aug))

print("\n--- FINAL RESULTS ---")

print("Baseline Accuracy:", np.mean(acc_real))
print("Baseline F1:", np.mean(f1_real))

print("Augmented Accuracy:", np.mean(acc_aug))
print("Augmented F1:", np.mean(f1_aug))


Fold 1
Epoch 1/30 - Loss: 99.7542
Epoch 10/30 - Loss: 82.8719
Epoch 20/30 - Loss: 77.8337
Epoch 30/30 - Loss: 70.5320
Epoch 1/30 - Loss: 103.3788
Epoch 10/30 - Loss: 84.6865
Epoch 20/30 - Loss: 80.9535
Epoch 30/30 - Loss: 79.3274
Epoch 1/30 - Loss: 87.8782
Epoch 10/30 - Loss: 81.3097
Epoch 20/30 - Loss: 73.9892
Epoch 30/30 - Loss: 63.0945

Fold 2
Epoch 1/30 - Loss: 96.2289
Epoch 10/30 - Loss: 83.7282
Epoch 20/30 - Loss: 81.3205
Epoch 30/30 - Loss: 78.0980
Epoch 1/30 - Loss: 94.4722
Epoch 10/30 - Loss: 83.1744
Epoch 20/30 - Loss: 79.3883
Epoch 30/30 - Loss: 75.8829
Epoch 1/30 - Loss: 96.2533
Epoch 10/30 - Loss: 82.9180
Epoch 20/30 - Loss: 78.4356
Epoch 30/30 - Loss: 73.2334

Fold 3
Epoch 1/30 - Loss: 89.0340
Epoch 10/30 - Loss: 81.7014
Epoch 20/30 - Loss: 78.4855
Epoch 30/30 - Loss: 70.0417
Epoch 1/30 - Loss: 88.3434
Epoch 10/30 - Loss: 82.1165
Epoch 20/30 - Loss: 77.1376
Epoch 30/30 - Loss: 70.7200
Epoch 1/30 - Loss: 96.6131
Epoch 10/30 - Loss: 82.6112
Epoch 20/30 - Loss: 80.2256
Epoc

# Condtion by Condtion

In [179]:
def augment_training_data_ablation(
    X_train, y_train, metadata_train,
    vae_models,
    target_condition,
    latent_dim=5,
    n_fake=100
):

    X_aug = [X_train]
    y_aug = [y_train]

    mask = metadata_train["condition"].values == target_condition

    if mask.sum() < 10:
        return np.vstack(X_aug), np.hstack(y_aug)

    decoder = vae_models[target_condition]["decoder"]
    scaler = vae_models[target_condition]["scaler"]

    for label in [0, 1]:

        z = np.random.normal(size=(n_fake, latent_dim)).astype("float32")

        X_fake_scaled = decoder.predict(z, verbose=0)
        X_fake = scaler.inverse_transform(X_fake_scaled)

        y_fake = np.full(n_fake, label)

        X_aug.append(X_fake)
        y_aug.append(y_fake)

    return np.vstack(X_aug), np.hstack(y_aug)

In [180]:
conditions = ["P300", "eyesOpen", "eyesClosed"]

# store results across folds
all_ablation_results = []

gkf = GroupKFold(n_splits=5)

# --- Cross Validation Loop --- #
for fold, (train_idx, test_idx) in enumerate(gkf.split(X, y, groups)):

    print(f"\n--- Fold {fold+1} ---")

    # Split Data
    X_train, X_test = X[train_idx], X[test_idx]
    y_train, y_test = y[train_idx], y[test_idx]
    metadata_train = metadata_df.iloc[train_idx]

    # Train VAE
    vae_models = {}

    for cond in conditions:

        mask = metadata_train["condition"].values == cond
        X_cond = X_train[mask]

        if len(X_cond) < 10:
            continue

        encoder, decoder, scaler, _ = train_vae(
            X_cond,
            latent_dim=8,
            epochs=30
        )

        vae_models[cond] = {
            "decoder": decoder,
            "scaler": scaler
        }

    # to store results
    ablation_results = {}

    # conditon by condition
    for cond in conditions:

        # one condition only augmentation
        X_aug, y_aug = augment_training_data_ablation(
            X_train,
            y_train,
            metadata_train,
            vae_models,
            target_condition=cond,
            latent_dim=8,
            n_fake=20
        )

        clf = Pipeline([
            ("scaler", StandardScaler()),
            ("logreg", LogisticRegression(max_iter=1000))
        ])

        clf.fit(X_aug, y_aug)

        # evaluation 
        y_pred = clf.predict(X_test)

        acc = accuracy_score(y_test, y_pred)
        f1 = f1_score(y_test, y_pred)

        ablation_results[cond] = {
            "acc": acc,
            "f1": f1
        }

        print(f"{cond} | Acc: {acc:.3f} | F1: {f1:.3f}")

    # store fold results
    all_ablation_results.append(ablation_results)

rows = []

for fold_idx, fold_results in enumerate(all_ablation_results):
    for cond, metrics in fold_results.items():
        rows.append({
            "fold": fold_idx,
            "condition": cond,
            "accuracy": metrics["acc"],
            "f1": metrics["f1"]
        })

df = pd.DataFrame(rows)

summary = df.groupby("condition")[["accuracy", "f1"]].mean()

print("\n--- FINAL ABALATION RESULTS ---")
print(summary)


--- Fold 1 ---
Epoch 1/30 - Loss: 93.1607
Epoch 10/30 - Loss: 82.2434
Epoch 20/30 - Loss: 78.9410
Epoch 30/30 - Loss: 73.7503
Epoch 1/30 - Loss: 98.3528
Epoch 10/30 - Loss: 83.8755
Epoch 20/30 - Loss: 80.8713
Epoch 30/30 - Loss: 77.7716
Epoch 1/30 - Loss: 91.6497
Epoch 10/30 - Loss: 82.8103
Epoch 20/30 - Loss: 79.5922
Epoch 30/30 - Loss: 75.1680
P300 | Acc: 0.757 | F1: 0.710
eyesOpen | Acc: 0.757 | F1: 0.690
eyesClosed | Acc: 0.730 | F1: 0.667

--- Fold 2 ---
Epoch 1/30 - Loss: 99.5194
Epoch 10/30 - Loss: 84.0531
Epoch 20/30 - Loss: 80.7815
Epoch 30/30 - Loss: 75.6969
Epoch 1/30 - Loss: 99.4323
Epoch 10/30 - Loss: 82.9104
Epoch 20/30 - Loss: 80.4428
Epoch 30/30 - Loss: 76.4972
Epoch 1/30 - Loss: 92.2124
Epoch 10/30 - Loss: 82.6782
Epoch 20/30 - Loss: 79.4107
Epoch 30/30 - Loss: 72.9560
P300 | Acc: 0.703 | F1: 0.756
eyesOpen | Acc: 0.676 | F1: 0.727
eyesClosed | Acc: 0.676 | F1: 0.727

--- Fold 3 ---
Epoch 1/30 - Loss: 90.9711
Epoch 10/30 - Loss: 83.1997
Epoch 20/30 - Loss: 80.4519
Epo

In [181]:
rows = []

# baseline
rows.append({
    "method": "Baseline",
    "accuracy": np.mean(acc_real),
    "f1": np.mean(f1_real)
})

# full augmentation
rows.append({
    "method": "Full Augmentation",
    "accuracy": np.mean(acc_aug),
    "f1": np.mean(f1_aug)
})

# ablations
for cond in conditions:
    rows.append({
        "method": f"{cond}-only",
        "accuracy": summary.loc[cond, "accuracy"],
        "f1": summary.loc[cond, "f1"]
    })
df_all = pd.DataFrame(rows)

df_all["accuracy"] = df_all["accuracy"].round(3)
df_all["f1"] = df_all["f1"].round(3)
print(df_all)

              method  accuracy     f1
0           Baseline     0.763  0.772
1  Full Augmentation     0.768  0.781
2          P300-only     0.757  0.774
3      eyesOpen-only     0.757  0.767
4    eyesClosed-only     0.763  0.772


# Conditional VAE

In [120]:
from sklearn.preprocessing import OneHotEncoder

In [119]:
def build_encoder_c(input_dim, latent_dim, n_conditions, n_labels):
    x_in = layers.Input(shape=(input_dim,))
    c_in = layers.Input(shape=(n_conditions,))
    y_in = layers.Input(shape=(n_labels,))

    h = layers.Concatenate()([x_in, c_in, y_in])

    h = layers.Dense(64, activation="relu")(h)
    h = layers.Dense(32, activation="relu")(h)

    z_mean = layers.Dense(latent_dim)(h)
    z_log_var = layers.Dense(latent_dim)(h)

    return models.Model([x_in, c_in, y_in], [z_mean, z_log_var])

def build_decoder_c(output_dim, latent_dim, n_conditions, n_labels):
    z_in = layers.Input(shape=(latent_dim,))
    c_in = layers.Input(shape=(n_conditions,))
    y_in = layers.Input(shape=(n_labels,))

    h = layers.Concatenate()([z_in, c_in, y_in])

    h = layers.Dense(32, activation="relu")(h)
    h = layers.Dense(64, activation="relu")(h)

    out = layers.Dense(output_dim)(h)

    return models.Model([z_in, c_in, y_in], out)

class Sampling(layers.Layer):
    def call(self, inputs):
        z_mean, z_log_var = inputs
        epsilon = tf.random.normal(shape=tf.shape(z_mean))
        return z_mean + tf.exp(0.5 * z_log_var) * epsilon
    
def compute_loss(x, x_recon, z_mean, z_log_var):
    recon = tf.reduce_mean(tf.reduce_sum(tf.square(x - x_recon), axis=1))

    kl = -0.5 * tf.reduce_mean(
        tf.reduce_sum(
            1 + z_log_var - tf.square(z_mean) - tf.exp(z_log_var),
            axis=1
        )
    )

    return recon + kl

In [121]:
class CVAE(tf.keras.Model):
    def __init__(self, encoder, decoder, **kwargs):
        super().__init__(**kwargs)
        self.encoder = encoder
        self.decoder = decoder

    def train_step(self, data):
        x, c, y = data

        with tf.GradientTape() as tape:
            z_mean, z_log_var = self.encoder([x, c, y])
            z = Sampling()([z_mean, z_log_var])

            x_recon = self.decoder([z, c, y])

            loss = compute_loss(x, x_recon, z_mean, z_log_var)

        grads = tape.gradient(
            loss,
            self.encoder.trainable_weights + self.decoder.trainable_weights
        )

        self.optimizer.apply_gradients(
            zip(grads, self.encoder.trainable_weights + self.decoder.trainable_weights)
        )

        return {"loss": loss}

In [159]:
def train_cvae(X, cond, y,
               input_dim,
               latent_dim=8,
               epochs=30,
               n_conditions=3,
               n_labels=2):

    encoder = build_encoder_c(input_dim, latent_dim, n_conditions, n_labels)
    decoder = build_decoder_c(input_dim, latent_dim, n_conditions, n_labels)

    cvae = CVAE(encoder, decoder)
    cvae.compile(optimizer=tf.keras.optimizers.Adam())

    dataset = tf.data.Dataset.from_tensor_slices((X, cond, y)).batch(64)

    losses = []

    for epoch in range(epochs):
        epoch_losses = []

        for batch in dataset:
            logs = cvae.train_step(batch)
            epoch_losses.append(logs["loss"].numpy())

        avg_loss = np.mean(epoch_losses)
        losses.append(avg_loss)

        if (epoch + 1) % 10 == 0 or epoch == 0 or epoch == epochs - 1:
            print(f"Epoch {epoch+1}/{epochs} - Loss: {avg_loss:.4f}")

    return encoder, decoder, losses

In [161]:
# Augmentation

def generate_samples(decoder, n_samples, latent_dim, cond_vec, label_vec):

    z = np.random.normal(size=(n_samples, latent_dim)).astype("float32")

    C = np.repeat(cond_vec[np.newaxis, :], n_samples, axis=0)
    Y = np.repeat(label_vec[np.newaxis, :], n_samples, axis=0)

    return decoder.predict([z, C, Y], verbose=0)


def augment_condition_only(
    X, y, cond_raw,
    decoder,
    cond_encoder,
    label_encoder,
    latent_dim=8,
    n_fake=20
):

    X_aug = [X]
    y_aug = [y]

    for cond_class in cond_encoder.categories_[0]:

        cond_vec = cond_encoder.transform([[cond_class]])[0]

        for label in np.unique(y):

            y_vec = label_encoder.transform([[label]])[0]

            z = np.random.normal(size=(n_fake, latent_dim)).astype("float32")

            C = np.repeat(cond_vec[np.newaxis, :], n_fake, axis=0)
            Y = np.repeat(y_vec[np.newaxis, :], n_fake, axis=0)

            X_fake = decoder.predict([z, C, Y], verbose=0)

            y_fake = np.full(n_fake, label)

            X_aug.append(X_fake)
            y_aug.append(y_fake)

    return np.vstack(X_aug), np.hstack(y_aug)

In [162]:
gkf = GroupKFold(n_splits=5)

acc_real, f1_real = [], []
acc_aug, f1_aug = [], []

all_losses = []

for fold, (train_idx, test_idx) in enumerate(gkf.split(X, y, groups)):

    print(f"\nFOLD {fold+1}")

    X_train, X_test = X[train_idx], X[test_idx]
    y_train, y_test = y[train_idx], y[test_idx]
    meta_train = metadata_df.iloc[train_idx]

    # one-hot
    cond_encoder = OneHotEncoder(sparse_output=False)
    label_encoder = OneHotEncoder(sparse_output=False)

    C_train = cond_encoder.fit_transform(
        meta_train["condition"].values.reshape(-1, 1)
    )

    Y_train = label_encoder.fit_transform(
        y_train.reshape(-1, 1)
    )

    # train cvae
    encoder, decoder, losses = train_cvae(
        X_train.astype("float32"),
        C_train.astype("float32"),
        Y_train.astype("float32"),
        input_dim=X_train.shape[1]
    )

    all_losses.append(losses)

    # baseline
    clf = Pipeline([
        ("scaler", StandardScaler()),
        ("logreg", LogisticRegression(max_iter=1000))
    ])

    clf.fit(X_train, y_train)
    pred = clf.predict(X_test)

    acc_real.append(accuracy_score(y_test, pred))
    f1_real.append(f1_score(y_test, pred))

    # augmentation
    X_aug, y_aug = augment_condition_only(
        X_train,
        y_train,
        meta_train["condition"].values,
        decoder,
        cond_encoder,
        label_encoder
    )

    clf_aug = Pipeline([
        ("scaler", StandardScaler()),
        ("logreg", LogisticRegression(max_iter=1000))
    ])

    clf_aug.fit(X_aug, y_aug)
    pred_aug = clf_aug.predict(X_test)

    acc_aug.append(accuracy_score(y_test, pred_aug))
    f1_aug.append(f1_score(y_test, pred_aug))

    print(f"Baseline F1: {f1_real[-1]:.3f} | Aug F1: {f1_aug[-1]:.3f}")


FOLD 1
Epoch 1/30 - Loss: 5.1223
Epoch 10/30 - Loss: 1.0980
Epoch 20/30 - Loss: 0.5503
Epoch 30/30 - Loss: 0.3798
Baseline F1: 0.667 | Aug F1: 0.839

FOLD 2
Epoch 1/30 - Loss: 4.6030
Epoch 10/30 - Loss: 1.0891
Epoch 20/30 - Loss: 0.6664
Epoch 30/30 - Loss: 0.4198
Baseline F1: 0.727 | Aug F1: 0.698

FOLD 3
Epoch 1/30 - Loss: 6.0329
Epoch 10/30 - Loss: 1.3064
Epoch 20/30 - Loss: 0.6532
Epoch 30/30 - Loss: 0.4444
Baseline F1: 0.829 | Aug F1: 0.800

FOLD 4
Epoch 1/30 - Loss: 4.3623
Epoch 10/30 - Loss: 0.8978
Epoch 20/30 - Loss: 0.5387
Epoch 30/30 - Loss: 0.3657
Baseline F1: 0.703 | Aug F1: 0.706

FOLD 5
Epoch 1/30 - Loss: 4.3040
Epoch 10/30 - Loss: 0.8374
Epoch 20/30 - Loss: 0.4860
Epoch 30/30 - Loss: 0.3254
Baseline F1: 0.933 | Aug F1: 0.930


In [163]:
print("\n--- FINAL RESULTS ---")

print("Baseline Accuracy:", np.mean(acc_real))
print("Baseline F1:", np.mean(f1_real))

print("Augmented Accuracy:", np.mean(acc_aug))
print("Augmented F1:", np.mean(f1_aug))


--- FINAL RESULTS ---
Baseline Accuracy: 0.7626683826683827
Baseline F1: 0.7718487445316713
Augmented Accuracy: 0.7788931788931789
Augmented F1: 0.7944998014209436


The application of generative augmentation revealed a clear distinction between unstructured and conditional modeling approaches. The baseline classifier achieved an accuracy of approximately 0.763 and an F1-score of 0.772, establishing a strong reference point. Augmentation using a standard VAE yielded negligible improvements in accuracy (~0.1%) and only marginal gains in F1 (~1%), suggesting that unconditioned latent sampling largely reproduces the global EEG distribution without enhancing class-discriminative features. In contrast, the CVAE framework, which explicitly incorporates both condition and label information during generation, produced more consistent benefits, with accuracy increasing by approximately 1% and F1 by 2%. Notably, the improvement was more pronounced in F1-score than in accuracy, indicating that conditional augmentation primarily enhanced the classifier’s ability to balance precision and recall, likely by improving representation of the minority or harder-to-classify class. However, the overall magnitude of improvement remained modest, reflecting the inherent difficulty of modeling EEG data, which is characterized by high noise, non-stationarity, and substantial inter-subject variability. These findings suggest that while conditional generative models can provide targeted benefits by aligning synthetic data with task-relevant structure, their impact is constrained by the quality of the learned latent space and the limited capacity of linear classifiers to exploit subtle distributional shifts. Overall, the results support the conclusion that CVAE-based augmentation offers a meaningful but incremental improvement over both baseline and unconditioned VAE approaches in EEG classification tasks.

Future Directions

1. Adopt stronger classifiers: Replace the current linear model with more expressive approaches (e.g., support vector machines, gradient boosting, or shallow neural networks) to better capture nonlinear relationships and leverage synthetic data.

2. Develop CNN-based CVAE architectures: Extend the current dense CVAE to a convolutional framework to better model the spatial and temporal structure of EEG signals, potentially improving the realism and utility of generated samples.